### PR TITLE
Proposed notes for PR Review on Jul 17th (PR #15169).

### PR DESCRIPTION
**Pull request by jachiang**

### CODE REVIEW

File: _posts/2019-07-17-#15169.md [hunk 1]
```diff
@@ -7,7 +7,30 @@ title: "#15169 Parallelize CheckInputs() in AcceptToMemoryPool()"
...
+  * Script check worker loop (Currently used for Block validation): https://github.com/bitcoin/bitcoin/blob/master/src/checkqueue.h#L66 
+* There is a significant performance gain resulting from this optimization.
+* This PR also changes behavior when transactions are  denied mempool acceptance. Unlike `CheckInputs`, `RunCheckInputsMaybeParallel` can no longer set TX_NOT_STANDARD during input check failure, but only a consensus failure state.
+  * Note: This has an effect on peer connections. Peers are not dropped by a node when it receives a non-standard but consensus-valid transaction. Instead, a peer will only be dropped when consensus-invalid transactions are received.
+
 ## Questions
 
+* When does `RunCheckInputsMaybeParallel()` still potentially call `CheckInputs()`?
+* How are the input check jobs dispatched to the thread pool? How are the check results returned after execution?
+* Does `RunCheckInputsMaybeParallel()` set the same "invalid reason" on input check failure as CheckInputs()?
```
Comment (jonatack) [inline _posts/2019-07-17-#15169.md]: nit: perhaps CheckInputs() enclosed in backticks too?

File: _posts/2019-07-17-#15169.md [hunk 2]
```diff
@@ -7,7 +7,30 @@ title: "#15169 Parallelize CheckInputs() in AcceptToMemoryPool()"
...
+    - sequentially and synchronously, in which case `CheckInputs()` will return `true` if all scripts pass verification or pass back the error for the first script that failed verification; or
+    - in parallel, in which case `CheckInputs()` will emplace the script check onto a queue for future verification.
+* Previously, `CheckInputs()` would always check inputs sequentially when accepting a transaction to the mempool. Transaction inputs could be checked in parallel during block validation. Note the [`vChecks` argument when calling `CheckInputs()` from `ConnectBlock()`](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/validation.cpp#L1887).
+* This PR enables the input checks to be performed in parallel when a transaction is entering the mempool. 
+* Currently, `CheckInputs()` will validate the inputs of a new transaction in sequence, and is called updating the mempool with a new transaction. 
+* This PR replaces calls to `CheckInputs()` with `RunCheckInputsMaybeParallel()`, which will push the input checks to an existing CCheckQueue worker queue.
+  * `CheckInputs`: https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1234
```
Comment (jonatack) [inline _posts/2019-07-17-#15169.md]: Are the 3 items in L17-19 meant to be indented as sub-items? You do that with more indentation and with dashes in L11-12 above.

File: _posts/2019-07-17-#15169.md [hunk 3]
```diff
@@ -7,7 +7,30 @@ title: "#15169 Parallelize CheckInputs() in AcceptToMemoryPool()"
...
+  * Checkqueue (Currently used for Block validation): https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1528
```
Comment (jonatack) [inline _posts/2019-07-17-#15169.md]: Why is the bool CheckInputs in L17 in backticks and not Checkqueue in L18 (seems it should be CCheckQueue as well)

File: _posts/2019-07-17-#15169.md [hunk 4]
```diff
@@ -7,7 +7,30 @@ title: "#15169 Parallelize CheckInputs() in AcceptToMemoryPool()"
...
+* Previously, `CheckInputs()` would always check inputs sequentially when accepting a transaction to the mempool. Transaction inputs could be checked in parallel during block validation. Note the [`vChecks` argument when calling `CheckInputs()` from `ConnectBlock()`](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/validation.cpp#L1887).
+* This PR enables the input checks to be performed in parallel when a transaction is entering the mempool. 
+* Currently, `CheckInputs()` will validate the inputs of a new transaction in sequence, and is called updating the mempool with a new transaction. 
+* This PR replaces calls to `CheckInputs()` with `RunCheckInputsMaybeParallel()`, which will push the input checks to an existing CCheckQueue worker queue.
+  * `CheckInputs`: https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1234
+  * Checkqueue (Currently used for Block validation): https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1528 
+  * Script check worker loop (Currently used for Block validation): https://github.com/bitcoin/bitcoin/blob/master/src/checkqueue.h#L66
```
Comment (jonatack) [inline _posts/2019-07-17-#15169.md]: s/Block/block/ (this item seems unfinished)

File: _posts/2019-07-17-#15169.md (hunk 2 above)
Comment (jonatack) [reply _posts/2019-07-17-#15169.md]: Just saw that L11-12 were added by John; makes more sense. Question still open regarding L17-19 items or sub-items.

File: _posts/2019-07-17-#15169.md (hunk 3 above)
Comment (jachiang) [reply _posts/2019-07-17-#15169.md]: You are right, fixed. Thank you.

File: _posts/2019-07-17-#15169.md (hunk 2 above)
Comment (jachiang) [reply _posts/2019-07-17-#15169.md]: Fixed. Thanks a lot for reviewing!

File: _posts/2019-07-17-#15169.md (hunk 1 above)
Comment (jachiang) [reply _posts/2019-07-17-#15169.md]: Fixed.

File: _posts/2019-07-17-#15169.md (hunk 4 above)
Comment (jachiang) [reply _posts/2019-07-17-#15169.md]: Not sure what you mean?

Comment (jonatack) [reply _posts/2019-07-17-#15169.md]: I meant that "block" didn't need to be capitalized, but it's a nit not worth fixing unless there is more you want to change.

File: _posts/2019-07-17-#15169.md [hunk 5]
```diff
@@ -7,7 +7,29 @@ title: "#15169 Parallelize CheckInputs() in AcceptToMemoryPool()"
...
+  * [`CCheckQueue`](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/validation.cpp#L1528) (Currently used for Block validation)
+  * [Script check worker loop](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/checkqueue.h#L66) (Currently used for Block validation)
+* There is a significant performance gain resulting from this optimization.
+* This PR also changes behavior when transactions are denied mempool acceptance. `RunCheckInputsMaybeParallel` (and by extension [`CheckInputs()`](https://github.com/bitcoin/bitcoin/pull/15169/commits/5f4e514412fc39e174f298a7737eb9f08a82a86b#diff-24efdb00bfbe56b140fb006b562cc70bL1436)) no longer set TX_NOT_STANDARD during input check failure, but only a consensus failure state.
+  * Note: This has an effect on peer connections. Peers are not dropped by a node when it receives a non-standard but consensus-valid transaction. Instead, a peer will only be dropped when consensus-invalid transactions are received. Since input checks for new transactions
```
Comment (jonatack) [inline _posts/2019-07-17-#15169.md]: Is there text missing after "Since input checks for new transactions"...?

File: _posts/2019-07-17-#15169.md [hunk 6]
```diff
@@ -7,7 +7,29 @@ title: "#15169 Parallelize CheckInputs() in AcceptToMemoryPool()"
...
+    - sequentially and synchronously, in which case `CheckInputs()` will return `true` if all scripts pass verification or pass back the error for the first script that failed verification; or
+    - in parallel, in which case `CheckInputs()` will emplace the script check onto a queue for future verification.
+* Previously, `CheckInputs()` would always check inputs sequentially when accepting a transaction to the mempool. Transaction inputs are checked in parallel during block validation. Note the [`vChecks` argument when calling `CheckInputs()` from `ConnectBlock()`](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/validation.cpp#L1887).
+* This PR enables the input checks to be performed in parallel when a transaction is entering the mempool. 
+* It does so by replacing calls to `CheckInputs()` with [`RunCheckInputsMaybeParallel()`](https://github.com/bitcoin/bitcoin/commit/cb0c42cfda7669f3df57c167cf0d691f926039d2#diff-24efdb00bfbe56b140fb006b562cc70bR1679), which will push the input checks to an existing CCheckQueue worker queue.
```
Comment (jonatack) [inline _posts/2019-07-17-#15169.md]: Possibly CCheckQueue in backticks here too.

File: _posts/2019-07-17-#15169.md (hunk 5 above)
Comment (jachiang) [reply _posts/2019-07-17-#15169.md]: Thanks again. Not my finest hour.

File: _posts/2019-07-17-#15169.md [hunk 7]
```diff
@@ -7,7 +7,29 @@ title: "#15169 Parallelize CheckInputs() in AcceptToMemoryPool()"
 
 ## Notes
 
+* The [CheckInputs()](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/validation.cpp#L1234) function is called when checking the inputs to a transaction. Each input has a script which is verified. That verification can be run:
+    - sequentially and synchronously, in which case `CheckInputs()` will return `true` if all scripts pass verification or pass back the error for the first script that failed verification; or
+    - in parallel, in which case `CheckInputs()` will emplace the script check onto a queue for future verification.
+* Previously, `CheckInputs()` would always check inputs sequentially when accepting a transaction to the mempool. Transaction inputs are checked in parallel during block validation. Note the [`vChecks` argument when calling `CheckInputs()` from `ConnectBlock()`](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/validation.cpp#L1887).
```
Comment (jnewbery) [inline _posts/2019-07-17-#15169.md]: s/inputs are checked in parallel during block validation/inputs can be checked in parallel during block validation/

Whether they are depends on the `-par` command line argument. See https://github.com/bitcoin/bitcoin/blob/8f604361ebaa5263e614c21570a3256e4dbc3bcc/src/init.cpp#L401

File: _posts/2019-07-17-#15169.md [hunk 8]
```diff
@@ -7,7 +7,29 @@ title: "#15169 Parallelize CheckInputs() in AcceptToMemoryPool()"
...
+  * [`CheckInputs`](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/validation.cpp#L1234)
+  * [`CCheckQueue`](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/validation.cpp#L1528) (Currently used for block validation)
+  * [Script check worker loop](https://github.com/bitcoin/bitcoin/blob/8f9725c83f1da5364354d2820696c84d542d3d37/src/checkqueue.h#L66) (Currently used for block validation)
+* There is a significant performance gain resulting from this optimization.
+* This PR also changes behavior when transactions are denied mempool acceptance. `RunCheckInputsMaybeParallel` (and by extension [`CheckInputs()`](https://github.com/bitcoin/bitcoin/commit/5f4e514412fc39e174f298a7737eb9f08a82a86b#diff-24efdb00bfbe56b140fb006b562cc70bL1436)) no longer set TX_NOT_STANDARD during input check failure, but only a consensus failure state.
+  * Note: This has an effect on peer connections. Peers are not dropped by a node when it receives a non-standard but consensus-valid transaction. Instead, a peer will only be dropped when consensus-invalid transactions are received.
```
Comment (jnewbery) [inline _posts/2019-07-17-#15169.md]: I don't think this is quite accurate (or at least confusing in the way you've written it)

Current behaviour:

- we receive a tx that is policy-invalid but consensus valid: don't disconnect the peer
- we receive a tx that is consensus-invalid: disconnect the peer

Proposed behaviour:

- we receive a tx that is policy-invalid or consensus-invalid: don't disconnect the peer

So now we won't disconnect a peer that sends us a consensus-invalid tx.

File: _posts/2019-07-17-#15169.md [hunk 9]
```diff
@@ -7,7 +7,29 @@ title: "#15169 Parallelize CheckInputs() in AcceptToMemoryPool()"
...
+
 ## Questions
 
+* When does `RunCheckInputsMaybeParallel()` still potentially call `CheckInputs()`?
+* How are the input check jobs dispatched to the thread pool? How are the check results returned after execution?
+* Does `RunCheckInputsMaybeParallel()` set the same "invalid reason" on input check failure as `CheckInputs()`?
+* If not, why is this not possible?
+* How does this change affect RPC methods such as [`testmempoolaccept`](https://github.com/bitcoin/bitcoin/commit/cb0c42cfda7669f3df57c167cf0d691f926039d2#diff-01aa7d1d32f1b9e5a836c9c411978918L928)?
+* Given the input check benchmarks posted by Suhas in the PR comments, what vulnerability does this PR aim to fix?
```
Comment (jnewbery) [inline _posts/2019-07-17-#15169.md]: Please remove this. No need to draw attention here!

**Comment by jnewbery** [comment]
ACK 1148f37f2f91c8605ee9679a4fb42386cf78a525

Thanks James! I'll squash and merge.

**Pull request merged**